internal/tuning.go

```diff
@@ -0,0 +1,455 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+
+	"go.temporal.io/sdk/internal/common/metrics"
+	"go.temporal.io/sdk/log"
+)
+
+// WorkerTuner allows for the dynamic customization of some aspects of worker behavior.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type WorkerTuner interface {
+	// GetWorkflowTaskSlotSupplier returns the SlotSupplier used for workflow tasks.
+	GetWorkflowTaskSlotSupplier() SlotSupplier
+	// GetActivityTaskSlotSupplier returns the SlotSupplier used for activity tasks.
+	GetActivityTaskSlotSupplier() SlotSupplier
+	// GetLocalActivitySlotSupplier returns the SlotSupplier used for local activities.
+	GetLocalActivitySlotSupplier() SlotSupplier
+	// GetNexusSlotSupplier returns the SlotSupplier used for nexus tasks.
+	GetNexusSlotSupplier() SlotSupplier
+}
+
+// SlotPermit is a permit to use a slot.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type SlotPermit struct {
+	// UserData is a field that can be used to store arbitrary on a permit by SlotSupplier
+	// implementations.
+	UserData any
+	// Specifically eager activities need to keep track of their own concurrent max separately and
+	// this helps them do that. It can be used for other specific use cases in the future.
+	extraReleaseCallback func()
+}
+
+// SlotReservationInfo contains information that SlotSupplier instances can use during
+// reservation calls. It embeds a standard Context.
+type SlotReservationInfo interface {
```

Comment:
Should it also include worker type?

Reply:
Eh, Java doesn't have it either. The problem is there's no typed way to do that right now since they're just strings. Could make a brand new enum for it I suppose. I will add the other things Java has.

Reply:
I would be fine passing the string through I don't think it needs to be typed. Fine leaving stuff off here that we can add later. I just wasn't sure if it was intentional or not